utils/manifest-utils/tests/certificate_expiration.rs

```diff
@@ -0,0 +1,284 @@
+use utils::*;
+
+use ya_manifest_utils::Keystore;
+
+#[test]
+fn accept_not_expired_certificate() {
+    let test_cert_dir = tempfile::tempdir().unwrap();
+
+    let valid_from = chrono::Utc::now()
+        .checked_sub_signed(chrono::Duration::days(2))
+        .unwrap();
+
+    let valid_to = chrono::Utc::now()
+        .checked_add_signed(chrono::Duration::days(2))
+        .unwrap();
+
+    let (self_signed_cert, ca_key_pair) =
+        create_self_signed_certificate(valid_from, valid_to).unwrap();
+
+    write_cert_to_file(
+        &self_signed_cert,
+        &test_cert_dir.path().join("self_signed.pem"),
+    );
+
+    let sut = Keystore::load(&test_cert_dir).unwrap();
+
+    let (req, csr_key_pair) = create_csr().unwrap();
+    let signed_cert = sign_csr(req, csr_key_pair.clone(), &self_signed_cert, ca_key_pair).unwrap();
+
+    let data = "DEADFACE";
+    let sig_alg = "sha256".to_string();
+    let sig = sign_data(data.as_bytes(), &csr_key_pair, &sig_alg);
+
+    assert!(sut
+        .verify_signature(
+            base64::encode(signed_cert.to_pem().unwrap()),
+            base64::encode(sig),
+            sig_alg,
+            data
+        )
+        .is_ok());
+}
+
+#[test]
+fn not_accept_expired_certificate() {
+    let test_cert_dir = tempfile::tempdir().unwrap();
+
+    let valid_from = chrono::Utc::now()
+        .checked_sub_signed(chrono::Duration::days(2))
+        .unwrap();
+
+    let valid_to = chrono::Utc::now()
+        .checked_sub_signed(chrono::Duration::days(1))
+        .unwrap();
+
+    let (self_signed_cert, ca_key_pair) =
+        create_self_signed_certificate(valid_from, valid_to).unwrap();
+
+    write_cert_to_file(
+        &self_signed_cert,
+        &test_cert_dir.path().join("self_signed.pem"),
+    );
+
+    let sut = Keystore::load(&test_cert_dir).unwrap();
+
+    let (req, csr_key_pair) = create_csr().unwrap();
+    let signed_cert = sign_csr(req, csr_key_pair.clone(), &self_signed_cert, ca_key_pair).unwrap();
+
+    let data = "DEADFACE";
+    let sig_alg = "sha256".to_string();
+    let sig = sign_data(data.as_bytes(), &csr_key_pair, &sig_alg);
+
+    assert!(sut
+        .verify_signature(
+            base64::encode(signed_cert.to_pem().unwrap()),
+            base64::encode(sig),
+            sig_alg,
+            data
+        )
+        .is_err());
```

Comment:
The error message could contain messages like "Invalid certificate" "Unknown signature algorithm: {}" and "Invalid signature". 
Maybe we should expect "Invalid certificate" here?

Reply:
We have considered this profoundly and it seems like such approach with focus on error appearance might be more reliable. What in situation when error message changes? What other advantage can you see when focused on exact error message?

Reply:
You may end up with false positive in case certificate validation will work but only signature validation will fail.
To be fair if this Error would be an Enum type there would be no excuse, but I did not needed it to control the Error handling flow - I only needed a different message, so it is a String.
The only importance of this message is the fact it will appear in an Offer Rejection, which is a kind of important feedback for the Requestor, because it tells him whether theirs certificate is invalid, or they provided wrong algorithm, or the signature does not work, which is why I tested these messages in other tests (so the fact you do not have there is not a huge deal).

Also it is more of a review of my own code, but probably we should have some proper Error hierarchy where an Error reaching user in one form or another should be of a different type, but we do not have such hierarchy, and I had not introduced it either...